---
title: "Translation efficiency (TE)"
output: html_document
---

## Analysis of changes in translation efficiency (TE)

```{r}
library(data.table)
library(dplyr)
library(DESeq2)
library(rstatix)
library(rcompanion)
library(ggplot2)
library(ggpubr)
library(ggh4x)
source("../functions_misc.R")
```

DESeq
```{r}
# countData
tab <- read.csv("../../Data/RNAseq-Riboseq/pab1_strains_RSEM_expected_count_isoforms.results.csv")
rownames(tab) <- tab$transcript_id
tab <- tab[, colnames(tab)[!grepl(pattern = "^RPF.*", x = colnames(tab))]]
tab <- round(x = tab[, -1], digits = 0)
tab <- tab[rowSums(tab) > 0, ] # Remove rows with no reads in any samples

# colData
cdt <- data.frame(fullname = colnames(tab))
cdt$strain <- sub("^RNA_", "", cdt$fullname)
cdt$strain <- sub("^CDS_", "", cdt$strain)
cdt$strain <- sub("_rep[0-9]", "", cdt$strain)
cdt$type <- sub("_.*", "", cdt$fullname)
```
```{r}
dds <- DESeqDataSetFromMatrix(countData = tab, colData = cdt, design = ~strain + strain:type)
dds$strain <- relevel(dds$strain, ref = "WT")
dds$type <- relevel(dds$type, ref = "RNA")
DGE <- DESeq(dds)
```

results TE pairwise comparison
```{r}
resultsNames(DGE)
res_t <- list()
res_t[["pbp1\u0394 / WT"]] <- results(DGE, contrast = list(c("strainpbp1d.typeCDS"), c("strainWT.typeCDS")), pAdjustMethod = "fdr", alpha = 0.01, independentFiltering = TRUE, tidy = TRUE)
res_t[["pab1\u0394pbp1\u0394 / WT"]] <- results(DGE, contrast = list(c("strainpab1d_pbp1d.typeCDS"), c("strainWT.typeCDS")), pAdjustMethod = "fdr", alpha = 0.01, independentFiltering = TRUE, tidy = TRUE)
res_t[["pab1\u0394pbp1\u0394 / pbp1\u0394"]] <- results(DGE, contrast = list(c("strainpab1d_pbp1d.typeCDS"), c("strainpbp1d.typeCDS")), pAdjustMethod = "fdr", alpha = 0.01, independentFiltering = TRUE, tidy = TRUE)
openxlsx::write.xlsx(list('pbp1d vs WT' = res_t$`pbp1Δ / WT`, 'pab1d_pbp1d vs WT' = res_t$`pab1Δpbp1Δ / WT`, "pab1d_pbp1d vs pbp1d" = res_t$`pab1Δpbp1Δ / pbp1Δ`), file = "pab1_TE_DESeq2.xlsx")

res_t2 <- bind_rows(res_t, .id = "pair")
write.csv(res_t2, file = "pab1_TE_DESeq2.csv", row.names = FALSE, quote = FALSE)
```

### Features of "Up" and "Down" mRNAs from pab1Δpbp1Δ vs. pbp1Δ comparison

Define "Up" and "Down" mRNAs
```{r}
res_TE <- read.csv("pab1_TE_DESeq2.csv")
res_TE <- add_sig_fc(df = res_TE, pval_col = "padj", pval_cutoff = 0.05, log2FC_col = "log2FoldChange", log2FC_cutoff = 0,
                     plot_fc_cutoff = 5, plot_pval_cutoff = 10^-15)
res_TE$changes <- factor(res_TE$changes, levels = c("Up", "Down", "Unchanged"))
PPvP <- res_TE[which(res_TE$pair == "pab1Δpbp1Δ / pbp1Δ"), ]
te_up <- PPvP[which(PPvP$changes == "Up"), ]$row
te_down <- PPvP[which(PPvP$changes == "Down"), ]$row
```
```{r}
# Volcano plot
nn <- data.frame(table(pair = res_TE$pair, changes = res_TE$changes))
nn$ypos <- rep(c(10, 7.5, 5), each = 3)
p_TE_main <- plot_volcano(res = res_TE[which(res_TE$pair == "pab1Δpbp1Δ / pbp1Δ"), ], 
                          nn = nn[which(nn$pair == "pab1Δpbp1Δ / pbp1Δ"), ], xlabel = "in TE", 
                          point_label = NULL)
```

Gene ontology analysis of TE groups in pab1Δpbp1Δ / pbp1Δ
```{r}
de <- res_TE[which(res_TE$pair == "pab1Δpbp1Δ / pbp1Δ"), ]
de$mrna <- sub(".*_", "", de$row)
de$root <- sub("_.*", "", de$row)
de <- de[which(de$changes != "Unchanged" & de$mrna == "mRNA"), ]
de <- split(de, de$changes)
de <- lapply(de, function(x) x$root)
```
```{r}
library(gprofiler2)
gostres <- gost(query = de, organism = "scerevisiae")

# Export results as Excel
xx <- gostres$result
xx$parents <- as.character(xx$parents) # parents column is list. Convert to character before exporting
xx <- split(xx, xx$query)
openxlsx::write.xlsx(list('Up' = xx$Up, 'Down' = xx$Down), file = "pab1_TE_gene_ontology_pab1d-vs-pbp1d.xlsx")

# Plot results
gp <- gostplot(gostres, capped = TRUE, interactive = FALSE)
tohighlight <- gostres$result %>%
  group_by(query, source) %>%
  slice_min(order_by = p_value, n = 3, with_ties = FALSE)
gpp <- publish_gostplot(gp, highlight_terms = tohighlight, 
                        width = 8.5, height = 11, filename = "pab1_TE_gene_ontology_pab1d-vs-pbp1d.pdf")
```

Prepare table of mRNA features
```{r}
# Initiation features
feature_initiation <- read.csv("../../Data/mRNA features/mRNA_features_initiation.csv")

# Termination features
feature_termination <- read.csv("../../Data/mRNA features/mRNA_features_termination.csv")

# Readthrough efficiency
RE <- read.csv("../Readthrough/pab1_strains_readthrough_efficiency.csv")
RE <- RE[which(RE$strain == "pab1d"), c("transcript", "rpkm_cds", "rpkm_ext", "rte_ext")]

# poly(A) tracts
polyA <- read.csv("../../Data/mRNA features/polyA_tracts.csv")

# oligo(U)
oligoU <- read.csv("../../Data/mRNA features/oligoU.csv")

# Combine feature files
features <- left_join(feature_initiation, select(feature_termination, -c("l_tr", "l_utr5", "l_cds", "l_utr3", "random_factor", "random_num")), by = "transcript", suffix = c("", "_stop"))
features <- left_join(features, RE, by = "transcript")
features <- left_join(features, polyA, by = "transcript")
features <- left_join(features, oligoU, by = "transcript", suffix = c("", "_U"))
rm(feature_initiation, feature_termination, RE, polyA, oligoU)
features$TE_change_PPvsP <- "Reference"

# Join combined features to TE data
te_up_df <- features[which(features$transcript %in% te_up), ]
te_up_df$TE_change_PPvsP <- "Up"
te_down_df <- features[which(features$transcript %in% te_down), ]
te_down_df$TE_change_PPvsP <- "Down"
ud <- rbind(te_up_df, te_down_df)
ud <- rbind(ud, features)
ud$TE_change_PPvsP <- factor(ud$TE_change_PPvsP, levels = c("Reference", "Up", "Down"))
ud <- left_join(ud, PPvP[, c("row", "baseMean", "log2FoldChange")], by = c("transcript" = "row"))
rm(te_up_df, te_down_df, features)

genes_with_UTRs <- scan("../../Data/reference_set_2693_genes.txt", character())
udf <- ud[which(ud$transcript %in% genes_with_UTRs), ]
udf <- udf[which(udf$l_utr5 > 18 & udf$l_utr3 > 16), ] # This is limit in 5'-UTR and 3'-UTR lengths is for exploring nucleotide identity surrounding start / stop such that these nucleotides do not include the start / stop codons themselves and CDS positions
```

#### mRNA abundance
```{r}
# log2 fold change in mRNA abundance
res_rna <- read.csv("../DE RNA-seq/pab1_RNAseq_DESeq2.csv")
ud2 <- left_join(ud, res_rna[which(res_rna$pair == "pab1Δpbp1Δ / pbp1Δ"), ], by = c("transcript" = "row"), suffix = c("", "_rna"))
ud3 <- ud2[which(!(is.na(ud2$padj))), c("transcript", "TE_change_PPvsP", "log2FoldChange_rna")] # Filter out mRNAs with read count too low

# RPKM mRNA abundance
rna <- read.csv("../../Data/RNAseq-Riboseq/pab1_strains_RSEM_FPKM_isoforms.results.csv")
rownames(rna) <- rna$transcript_id
rna <- rna[, colnames(rna)[grepl(pattern = "^RNA.*", x = colnames(rna))]]
pbp1d_rpkm <- data.frame(transcript = rownames(rna), pbp1d_ave_RPKM = rowMeans(log10(rna[, 4:6])))
pbp1d_rpkm[which(pbp1d_rpkm$pbp1d_ave_RPKM == -Inf), ]$pbp1d_ave_RPKM <- -1.6 # Set -Inf to the closest lowest negative number
ud3 <- left_join(ud3, pbp1d_rpkm, by = "transcript")
```

Plot TE group vs mRNA abundance in WT condition (pbp1d cells)
```{r}
nn2 <- data.frame(table(TE_change_PPvsP = ud3$TE_change_PPvsP))
comps <- get_comparisons(ud3, "TE_change_PPvsP")
stat.test2 <- ud3 %>%
  wilcox_test(pbp1d_ave_RPKM~TE_change_PPvsP, p.adjust.method = "BH", paired = FALSE, comparisons = comps) %>%
  add_significance() %>% add_xy_position(step.increase = 0.3)
stat.test2$y.position <- c(3.75, 3.875, 4)
p_TE_pbp1d_mRNA <- ggplot(ud3, aes(x = TE_change_PPvsP, y = pbp1d_ave_RPKM)) + 
  geom_boxplot(aes(fill = TE_change_PPvsP), outlier.shape = NA) + 
  geom_text(data = nn2, aes(x = TE_change_PPvsP, y = -0.5, label = paste0("n = ", Freq)), size = 7/.pt) +
  stat_pvalue_manual(data = stat.test2, label = "p.adj.signif", tip.length = 0.001, size = 7/.pt, vjust = 0.5, hide.ns = TRUE) +
  scale_fill_manual(values = c(Up = "orange", Down = "purple", Reference = "white")) +
  scale_x_discrete(labels = c(Reference = "Ref.", Up = "Up", Down = "Down")) + 
  ylab(expression(atop("mRNA abundance", italic("pbp1Δ")))) +
  xlab(expression(atop("log"[2]*" fold change in TE", italic("pab1Δpbp1Δ / pbp1Δ")))) +
  coord_cartesian(ylim = c(-0.5, 4)) +
  theme_bw(base_size = 10) + 
  theme(strip.background = element_rect(fill = "white"), panel.grid = element_blank(), axis.title.x = element_text(face = "italic"), legend.position = "none")
```

Plot TE group vs mRNA abundance changes
```{r}
nn3 <- data.frame(table(TE_change_PPvsP = ud3$TE_change_PPvsP))
comps <- get_comparisons(ud3, "TE_change_PPvsP")
stat.test3 <- ud3 %>%
  wilcox_test(log2FoldChange_rna~TE_change_PPvsP, p.adjust.method = "BH", paired = FALSE, comparisons = comps) %>%
  add_significance() %>% add_xy_position(step.increase = 0.3)
stat.test3$y.position <- c(2.5, 2.75, 3)
p_TE_mRNA_changes <- ggplot(ud3, aes(x = TE_change_PPvsP, y = log2FoldChange_rna)) + 
  geom_boxplot(aes(fill = TE_change_PPvsP), outlier.shape = NA) + 
  geom_text(data = nn3, aes(x = TE_change_PPvsP, y = -2.5, label = paste0("n = ", Freq)), size = 7/.pt) +
  stat_pvalue_manual(data = stat.test3, label = "p.adj.signif", tip.length = 0.001, size = 7/.pt, vjust = 0.5, hide.ns = TRUE) +
  scale_fill_manual(values = c(Up = "orange", Down = "purple", Reference = "white")) +
  scale_x_discrete(labels = c(Reference = "Ref.", Up = "Up", Down = "Down")) + 
  ylab(expression(atop("log"[2]*" fold change in mRNA", italic("pab1Δpbp1Δ / pbp1Δ")))) +
  xlab(expression(atop("log"[2]*" fold change in TE", italic("pab1Δpbp1Δ / pbp1Δ")))) +
  coord_cartesian(ylim = c(-2.5, 3)) +
  theme_bw(base_size = 10) + 
  theme(strip.background = element_rect(fill = "white"), panel.grid = element_blank(), axis.title.x = element_text(face = "italic"), legend.position = "none")
```

#### eIF4G

Compare to eIF4G depletion ribo-seq data from Zinshteyn et al. 2017
```{r}
zins <- list()
# YPA media
zins$YPA <- read.table("../../Data/Zinshteyn_2017_GSE87614_YPA_TEs.tsv", header = TRUE, sep = "\t")
zins$YPA$gene <- sub("$", "_mRNA", zins$YPA$gene)
zins$YPA$YPA_TE_foldchange_2hr <- zins$YPA$YPA_4Gts_2h.TE / zins$YPA$YPA_WT_2h.TE
zins$YPA$YPA_TE_foldchange_0hr <- zins$YPA$YPA_4Gts_0h.TE / zins$YPA$YPA_WT_0h.TE
zins$YPA$YPA_TE_foldchange_preshift <- zins$YPA$YPA_4Gts_pre_shift.TE / zins$YPA$YPA_WT_pre_shift.TE
```
```{r}
comb <- left_join(ud[, c("transcript", "log2FoldChange", "TE_change_PPvsP")], zins$YPA[, c("gene", "YPA_TE_foldchange_2hr", "YPA_TE_foldchange_0hr", "YPA_TE_foldchange_preshift")], by = c("transcript" = "gene"))
comb_m <- reshape2::melt(comb)
comb_m_filter <- comb_m[comb_m$variable != "log2FoldChange", ]
comb_m_filter$facet <- sub("_TE_foldchange_", "", comb_m_filter$variable)
comb_m_filter$facet <- sub("YPA", "", comb_m_filter$facet)
comb_m_filter <- comb_m_filter[!is.na(comb_m_filter$value), ]
comb_m_filter$log2FC <- log2(comb_m_filter$value)
comb_m_filter <- comb_m_filter[is.finite(comb_m_filter$log2FC), ]
comb_m_filter$facet <- factor(comb_m_filter$facet, levels = c("preshift", "0hr", "2hr"))

comp <- get_comparisons(comb_m_filter, variable = "TE_change_PPvsP")
stat.test4 <- comb_m_filter %>% group_by(facet) %>% 
  wilcox_test(log2FC ~ TE_change_PPvsP, comparisons = comp, p.adjust.method = "BH")
stat.test4 <- stat.test4 %>% add_significance() %>% add_xy_position()
stat.test4$y.position <- rep(c(1.25, 1.375, 1.5), times = 3)

nnz <- data.frame(table(TE_change_PPvsP = comb_m_filter$TE_change_PPvsP, facet = comb_m_filter$facet))
p_riboseq <- ggplot(comb_m_filter, aes(x = TE_change_PPvsP, y = log2FC)) + 
  geom_boxplot(outlier.alpha = 0, aes(fill = TE_change_PPvsP), show.legend = FALSE) + 
  geom_text(data = nnz, aes(x = TE_change_PPvsP, y = -1.75, label = paste0("n = ", Freq)), size = 7/.pt) +
  stat_pvalue_manual(data = stat.test4, label = "p.adj.signif", tip.length = 0.002, size = 7/.pt, vjust = 0.5, hide.ns = TRUE) +
  facet_wrap(~facet, nrow = 1) +
  scale_fill_manual(values = c(Up = "orange", Down = "purple", Reference = "white")) +
  scale_x_discrete(labels = c(Reference = "Ref.", Up = "Up", Down = "Down")) + 
  #xlab(bquote(log[2] ~ "fold change in TE")) + 
  xlab(expression(atop(log[2] ~ "fold change in TE", italic("pab1Δpbp1Δ / pbp1Δ")))) +
  ylab(expression(atop(log[2] ~ "fold change in TE", italic("eIF4Gd / WT")))) +
  coord_cartesian(ylim = c(-1.75, 1.5)) +
  theme_bw(base_size = 10) + 
  theme(strip.background = element_rect(fill = "white"), panel.grid = element_blank(), axis.title.x = element_text(face = "italic"), legend.position = "none")
```

Compare to RIP-seq data from Costello et al. 2015
```{r}
costello2 <- list()
sheets <- c("eIF4G1", "eIF4G2", "Pab1p")
for (i in sheets) {
  costello2[[i]] <- as.data.frame(readxl::read_xlsx("../../Data/Costello_2015_13059_2014_559_MOESM2_ESM.xlsx", sheet = i))
  costello2[[i]]$genes <- sub("$", "_mRNA", costello2[[i]]$genes)
}

analyze_costello <- function(ud, IP_data = "eIF4G1", cols = c("transcript", "l_tr", "l_cds", "l_utr5", "l_utr3", "rpkm_cds", "rpkm_ext", "rte_ext", "TE_change_PPvsP", "baseMean", "log2FoldChange")) {
  udf_c2 <- left_join(ud[, cols], costello2[[IP_data]], by = c("transcript" = "genes"))
  udf_c2 <- udf_c2[which(!(is.na(udf_c2$logFC)) & !(is.na(udf_c2$log2FoldChange))), ]
  udf_c2$dir <- "Unchanged"
  udf_c2[which(udf_c2$FDR < 0.05 & udf_c2$logFC > 0), ]$dir <- "Enriched in IP"
  udf_c2[which(udf_c2$FDR < 0.05 & udf_c2$logFC < 0), ]$dir <- "Depleted in IP"
  pp <- table(udf_c2$TE_change_PPvsP, udf_c2$dir)
  require(rcompanion)
  chi_res <- pairwiseNominalIndependence(pp, method = "BH")
  ppm <- as.data.frame(pp)
  ppm <- ppm %>% group_by(Var1) %>% mutate(percent = 100*Freq/sum(Freq))
  res_list <- list(df = udf_c2, chi_res = chi_res, freq_res = ppm)
  return(res_list)
}

res_ripseq <- sapply(names(costello2), function(x) analyze_costello(ud = ud, IP_data = x), simplify = FALSE, USE.NAMES = TRUE)
names(res_ripseq)[3] <- "Pab1" # Change Pab1p to Pab1
```
```{r}
# Plot frequencies
aa <- bind_rows(lapply(res_ripseq, function(x) x$freq_res), .id = "IP")
aa$Var2 <- factor(aa$Var2, levels = c("Depleted in IP", "Unchanged", "Enriched in IP"))
colnames(aa)[2:3] <- c("TE_change", "IP_change")
write.csv(aa, file = "../../Figures/data/RIPseq.csv", quote = FALSE, row.names = FALSE)

p_ripseq <- ggplot(aa, aes(x = TE_change, y = percent, fill = IP_change)) + 
  geom_col(position = "stack", color = "grey50") + 
  geom_text(aes(label = paste0(Freq, "\n(", round(percent), "%)")), 
            position = position_stack(vjust = 0.5), size = 6/.pt) +
  facet_wrap(~IP, nrow = 1) +
  scale_fill_manual(name = "RIP-seq: ", values = c("#F8766D", "grey50", "#00B0F6")) +
  scale_x_discrete(labels = c(Reference = "Ref.", Up = "Up", Down = "Down")) + 
  xlab("") + ylab("Percentage") +
  theme_bw(base_size = 10) + 
    theme(strip.background = element_rect(fill = "white"), panel.grid = element_blank(), 
          axis.title.x = element_blank(), legend.position = "top")
```
```{r}
# Plot logFC
colnames(PPvP)[c(2, 10)] <- c("transcript", "TE_change_PPvsP")
bb <- sapply(names(costello2), function(x) analyze_costello(ud = PPvP, IP_data = x, cols = c("transcript", "baseMean", "log2FoldChange", "sig", "TE_change_PPvsP")), simplify = FALSE, USE.NAMES = TRUE)
names(bb)[3] <- "Pab1" # Change Pab1p to Pab1
bb <- bind_rows(lapply(bb, function(x) x$df), .id = "IP")
bb$dir <- factor(bb$dir, levels = c("Depleted in IP", "Unchanged", "Enriched in IP"))
bb$TE_change_PPvsP <- factor(bb$TE_change_PPvsP, levels = c("Down", "Unchanged", "Up"))
p_ripseq_scatter <- ggplot(bb, aes(x = log2FoldChange, y = logFC)) + 
  geom_point(alpha = 0.3, shape = 21, aes(fill = TE_change_PPvsP, color = dir)) + 
  facet_grid(.~IP) + 
  stat_cor(method = "spearman", cor.coef.name = "rho", size = 7/.pt) + geom_smooth(method = "lm", color = "grey20") +
  geom_hline(yintercept = 0, linetype = "dashed") + geom_vline(xintercept = 0, linetype = "dashed") +
  scale_fill_manual(name = "TE changes: ", values = c(Up = "orange", Down = "purple", Unchanged = "grey50")) +
  scale_color_manual(name = "RIP-seq: ", values = c(`Depleted in IP` = "#F8766D", Unchanged = "grey50", `Enriched in IP` = "#00B0F6")) +
  #scale_shape_manual(name = "RIP-seq: ", values = c(`Depleted in IP` = 25, Unchanged = 1, `Enriched in IP` = 24)) +
  xlab(expression(atop(log[2] ~ "fold change in TE", italic("pab1Δpbp1Δ / pbp1Δ")))) +
  ylab(expression(log[2] ~ "fold change in RIP-seq")) +
  theme_bw(base_size = 10) + 
  theme(strip.background = element_rect(fill = "white"), panel.grid = element_blank(), legend.position = "top") +
  guides(colour = guide_legend(override.aes = list(alpha = 1)), fill = guide_legend(override.aes = list(alpha = 1)))
```
```{r}
comp <- get_comparisons(bb, variable = "TE_change_PPvsP")
stat.test5 <- bb %>% group_by(IP) %>% 
  wilcox_test(logFC ~ TE_change_PPvsP, comparisons = comp, p.adjust.method = "BH")
stat.test5 <- stat.test5 %>% add_significance() %>% add_xy_position()
stat.test5$y.position <- rep(c(2.5, 2.75, 3), times = 3)

nnr <- data.frame(table(TE_change_PPvsP = bb$TE_change_PPvsP, IP = bb$IP))
p_ripseq2 <- ggplot(bb, aes(x = TE_change_PPvsP, y = logFC)) +
  geom_boxplot(outlier.alpha = 0, aes(fill = TE_change_PPvsP), show.legend = FALSE) +
  geom_text(data = nnr, aes(x = TE_change_PPvsP, y = -2.75, label = paste0("n = ", Freq)), size = 7/.pt) +
  stat_pvalue_manual(data = stat.test5, label = "p.adj.signif", tip.length = 0.002, size = 7/.pt, vjust = 0.5, hide.ns = TRUE) +
  facet_wrap(~IP, nrow = 1) +
  scale_fill_manual(values = c(Up = "orange", Down = "purple", Reference = "white")) +
  scale_x_discrete(labels = c(Reference = "Ref.", Up = "Up", Down = "Down")) + 
  xlab(expression(atop(log[2] ~ "fold change in TE", italic("pab1Δpbp1Δ / pbp1Δ")))) +
  ylab(expression(log[2] ~ "fold change in RIP-seq")) +
  coord_cartesian(ylim = c(-3, 3)) +
  theme_bw(base_size = 10) + 
  theme(strip.background = element_rect(fill = "white"), panel.grid = element_blank(), axis.title.x = element_text(face = "italic"), legend.position = "none")
```
```{r}
# eIF4G sensitivity vs. transcript length
bbf <- bb[which(bb$transcript %in% udf$transcript), ]
bbf$TE_change_PPvsP <- recode_factor(bbf$TE_change_PPvsP, Reference = "Ref.", Up = "Up", Down = "Down")
bbf$dir <- recode_factor(bbf$dir, `Depleted in IP` = "Depleted", Unchanged = "Unchanged", `Enriched in IP` = "Enriched")

comp <- get_comparisons(bbf, variable = "dir")
stat.test6 <- bbf %>% group_by(IP, TE_change_PPvsP) %>% 
  wilcox_test(l_tr ~ dir, comparisons = comp, p.adjust.method = "BH")
stat.test6 <- stat.test6 %>% add_significance() %>% add_xy_position()
stat.test6$y.position <- rep(c(log10(6000), log10(7000), log10(8000)), times = 9)

nn6 <- data.frame(table(TE_change_PPvsP = bbf$TE_change_PPvsP, IP = bbf$IP, dir = bbf$dir))

p_ripseq3 <- ggplot(bbf, aes(x = dir, y = l_tr)) +
  geom_boxplot(outlier.alpha = 0, aes(fill = dir)) +
  geom_text(data = nn6, aes(x = dir, y = 200, label = Freq), size = 7/.pt, position = position_dodge(width = 0.2)) +
  stat_pvalue_manual(data = stat.test6, label = "p.adj.signif", tip.length = 0.002, size = 7/.pt, vjust = 0.5, hide.ns = TRUE) +
  facet_nested(.~IP+TE_change_PPvsP, strip = strip_split(c("bottom", "top"))) +
  scale_fill_manual(name = "RIP-seq: ", values = c("#F8766D", "grey50", "#00B0F6")) +
  scale_y_log10() +
  xlab("") + ylab("mRNA length (nt)") +
  theme_bw(base_size = 10) + 
  theme(strip.background = element_rect(fill = "white"), strip.placement = "outside",
        panel.grid = element_blank(), legend.position = "top",
        axis.title.x = element_text(face = "italic"), axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1))
```

#### Length of mRNA regions

```{r}
plot_compare_lengths <- function(udf, col_name = "l_cds", xlabel = "in TE", ylabel = "CDS length (nt)", n_ypos = 10, 
                                 y_log_transformation = "log10") {
  udf <- udf[, c("TE_change_PPvsP", col_name)]
  colnames(udf)[2] <- "l"
  if (y_log_transformation == "none") {
    udf$logl <- udf$l
  } else if (y_log_transformation == "log10") {
    udf$logl <- log10(udf$l) 
  } else {
    message("Invalid option")
  }
  comps <- get_comparisons(udf, "TE_change_PPvsP")
  stat.test <- udf %>%
    wilcox_test(logl~TE_change_PPvsP, p.adjust.method = "BH", paired = FALSE, comparisons = comps) %>%
    add_significance() %>% add_xy_position(step.increase = 0.3)
  nn <- data.frame(table(TE_change_PPvsP = udf$TE_change_PPvsP))
  p <- ggplot(udf, aes(x = TE_change_PPvsP, y = l)) + 
    geom_boxplot(outlier.alpha = 0.3, aes(fill = TE_change_PPvsP), show.legend = FALSE) + 
    geom_text(data = nn, aes(x = TE_change_PPvsP, y = n_ypos, label = paste0("n = ", Freq)), size = 7/.pt) +
    stat_pvalue_manual(data = stat.test, label = "p.adj.signif", tip.length = 0.01, size = 7/.pt, vjust = 0.5, hide.ns = TRUE) +
    scale_fill_manual(values = c(Up = "orange", Down = "purple", Reference = "white")) +
    scale_x_discrete(labels = c(Reference = "Ref.", Up = "Up", Down = "Down")) + 
    xlab(bquote(log[2] ~ "fold change" ~ .(xlabel))) + ylab(ylabel) +
    theme_bw(base_size = 10) + 
    theme(strip.background = element_rect(fill = "white"), panel.grid = element_blank(), axis.title.x = element_text(face = "italic"),
          legend.position = "none")
  if (y_log_transformation == "log10") {
    p <- p + scale_y_log10()
  }
  return(p)
}

p5 <- plot_compare_lengths(udf = udf, col_name = "l_utr5", xlabel = "in TE", ylabel = "5'-UTR length (nt)", n_ypos = 10)
pc <- plot_compare_lengths(udf = udf, col_name = "l_cds", xlabel = "in TE", ylabel = "CDS length (nt)", n_ypos = 50)
p3 <- plot_compare_lengths(udf = udf, col_name = "l_utr3", xlabel = "in TE", ylabel = "3'-UTR length (nt)", n_ypos = 10)
pm <- plot_compare_lengths(udf = udf, col_name = "l_tr", xlabel = "in TE", ylabel = "mRNA length (nt)", n_ypos = 200)
```

#### Readthrough

Detectable readthrough
```{r}
udff <- udf[which(udf$rpkm_cds > 2 & udf$rpkm_ext > 0.1), ] # 1490 mRNAs
udfrt <- table(udf$TE_change_PPvsP, udf$transcript %in% udff$transcript)
pairwise_fisher_test(xtab = udfrt, p.adjust.method = "BH", detailed = TRUE)

udfrt2 <- as.data.frame(udfrt)
udfrt2 <- udfrt2 %>% group_by(Var1) %>% mutate(percent = 100*Freq/sum(Freq))
udfrt2$Var2 <- recode_factor(udfrt2$Var2, `TRUE` = "Yes", `FALSE` = "No")
p_ref <- ggplot(udfrt2, aes(x = Var1, y = percent, fill = Var2)) + 
  geom_col(position = "stack") + 
  geom_text(aes(label = paste0(Freq, "\n(", round(percent), "%)")), 
            position = position_stack(vjust = 0.5), size = 7/.pt) +
  scale_fill_manual(name = "Readthrough detected: ", values = c("forestgreen", "grey50")) +
  scale_x_discrete(labels = c(Reference = "Ref.", Up = "Up", Down = "Down")) + 
  xlab("") + ylab("Percentage") +
  theme_bw(base_size = 10) + 
    theme(strip.background = element_rect(fill = "white"), panel.grid = element_blank(), 
          axis.title.x = element_blank(), legend.position = "top") +
  guides(fill = guide_legend(title.position = "top"))
```

Readthrough efficiency
```{r}
udff <- udf[which(udf$rpkm_cds > 2 & udf$rpkm_ext > 0.1), ] # 1490 mRNAs
udff$log_rte <- log2(udff$rte_ext)
p_re <- plot_compare_lengths(udf = udff, col_name = "log_rte", xlabel = "in TE", ylabel = "Readthrough efficiency", n_ypos = -12, y_log_transformation = "none")
```

#### Sequence motif

```{r}
analyze_motif <- function(udf, condition = list(udf$utr5_n_tracts > 0), facet_label = list("5'-UTR"), 
                          legend_label = "poly(A) tract present: ") {
  c_len <- length(condition)
  f_len <- length(facet_label)
  if (c_len == f_len) {
    fisher_res <- list()
    df_list <- list()
    for (i in 1:c_len) {
      udm <- table(TE_change = udf$TE_change_PPvsP, motif = condition[[i]])
      require(rstatix)
      fisher_res[[facet_label[[i]]]] <- pairwise_fisher_test(xtab = udm, p.adjust.method = "BH", detailed = TRUE)
      udm <- as.data.frame(udm)
      udm$motif <- recode_factor(udm$motif, `TRUE` = "Yes", `FALSE` = "No")
      udm <- udm %>% group_by(TE_change) %>% mutate(percent = 100*Freq/sum(Freq))
      df_list[[facet_label[[i]]]] <- udm
    }
    df <- bind_rows(df_list, .id = "facet_label")
    df$facet_label <- factor(df$facet_label, levels = unlist(facet_label))
    p <- ggplot(df, aes(x = TE_change, y = percent, fill = motif)) + 
      geom_col(position = "stack") + 
      geom_text(aes(label = paste0(Freq, "\n(", round(percent), "%)")), 
                position = position_stack(vjust = 0.5), size = 7/.pt) +
      facet_wrap(~facet_label) +
      scale_fill_manual(name = legend_label, values = c(Yes = "forestgreen", No = "grey50")) +
      scale_x_discrete(labels = c(Reference = "Ref.", Up = "Up", Down = "Down")) + 
      xlab("") + ylab("Percentage") +
      theme_bw(base_size = 10) + 
        theme(strip.background = element_rect(fill = "white"), panel.grid = element_blank(), 
              axis.title.x = element_blank(), legend.position = "top") +
      guides(fill = guide_legend(title.position = "top"))
  } else {
    message("Number of conditions and number of facet_label must match")
  }
  res_list <- list(fisher_res = fisher_res, plot_res = p, df_res = df)
  return(res_list)
}
```

poly(A) tracts in the UTRs
```{r}
# 5'-UTR
ud5_A <- analyze_motif(udf = udf, condition = list(udf$utr5_n_tracts > 0), facet_label = list("5'-UTR"), 
                       legend_label = "poly(A) tract present: ")
ud5_A$fisher_res

# 3'-UTR
ud3_A <- analyze_motif(udf = udf, condition = list(udf$utr3_n_tracts > 0), facet_label = list("3'-UTR"), 
                       legend_label = "poly(A) tract present: ")
ud3_A$fisher_res
```

oligo(U) in the 5'-UTR
```{r}
# 5'-UTR
ud5_U <- analyze_motif(udf = udf, condition = list(udf$utr5_n_tracts_U > 0), facet_label = list("5'-UTR"), 
                       legend_label = "oligo(U) present: ")
ud5_U$fisher_res
```

uORFs in the 5'-UTR
```{r}
uu <- colnames(udf)[grepl(pattern = "uORF", x = colnames(udf))]
udfu <- udf[, c("TE_change_PPvsP", "log2FoldChange", uu)]
udfu[is.na(udfu$uORF_count_overlap), ]$uORF_count_overlap <- 0
udfu[is.na(udfu$uORF_count_upstream), ]$uORF_count_upstream <- 0
udfu$uORF_count_all <- udfu$uORF_count_overlap + udfu$uORF_count_upstream

uorf <- analyze_motif(udf = udfu, condition = list(udfu$uORF_count_all > 0, udfu$uORF_count_upstream > 0, udfu$uORF_count_overlap > 0), 
                      facet_label = list("Any", "Upstream", "Overlapping"), 
                      legend_label = "uORF present: ")
```
```{r}
udfu2 <- udfu[which(udfu$TE_change_PPvsP == "Reference"), ]
udfu2_any <- udfu2
udfu2_any$uORF <- ifelse(udfu2_any$uORF_count_all > 0, yes = "Yes", no = "No")
udfu2_upstream <- udfu2
udfu2_upstream$uORF <- ifelse(udfu2_upstream$uORF_count_upstream > 0, yes = "Yes", no = "No")
udfu2_overlap <- udfu2
udfu2_overlap$uORF <- ifelse(udfu2_overlap$uORF_count_overlap > 0, yes = "Yes", no = "No")
udfu3 <- bind_rows(list(Any = udfu2_any, Upstream = udfu2_upstream, Overlapping = udfu2_overlap), .id = "type")

ks_results <- list()
for (x in c("Any", "Upstream", "Overlapping")) {
  ks_results[[x]] <- ks.test(x = udfu3[which(udfu3$type == x & udfu3$uORF == "Yes"), ]$log2FoldChange, 
                             y = udfu3[which(udfu3$type == x & udfu3$uORF == "No"), ]$log2FoldChange)$p.value
}
ks_results <- as.data.frame(unlist(ks_results))
ks_results$type <- row.names(ks_results)
colnames(ks_results)[1] <- "p_value"

uorf_cdf <- ggplot(udfu3, aes(x = log2FoldChange)) + 
  stat_ecdf(geom = "step", aes(color = uORF)) +
  geom_text(data = ks_results, aes(x = -2.5, y = 0.9, label = paste0('italic(p) == ', signif(p_value, digits = 3))), 
            size = 7/.pt, hjust = 0, parse = TRUE) +
  facet_wrap(~type, nrow = 1) +
  scale_color_manual(name = "uORF present: ", values = c(Yes = "forestgreen", No = "grey50"), limits = c("Yes", "No")) +
  xlab(expression(atop(log[2] ~ "fold change in TE", italic("pab1Δpbp1Δ / pbp1Δ")))) + 
  ylab("Cumulative Density") +
  theme_bw(base_size = 10) + 
  theme(strip.background = element_rect(fill = "white"), panel.grid = element_blank(), 
        legend.position = "top")
```


STREME results
```{r}
# https://meme-suite.org/meme/tools/streme
  # Select the type of control sequences to use: User-provided sequences 
  # Input sequences: 5'-UTR or 3'-UTR sequences of mRNAs in TE up or down group
  # Control sequences: 5'-UTR or 3'-UTR sequences of mRNAs in Reference
  # Advanced options: Minimum width = 3, Maximum width = 6 or 15, default for other parameters
library(ggseqlogo)
```
```{r}
# Motif enriched in 5'-UTR sequences of TE down group vs Reference
down5_res <- read.table("./STREME/streme_TE_down-vs-Ref_seq_utr5_3-6nt.txt", skip = 29, fill = TRUE, flush = TRUE)
down5 <- down5_res[3:7, 1:4]
down5 <- apply(down5, 1, as.numeric)
rownames(down5) <- c("A", "C", "G", "U")
pval <- as.numeric(down5_res[2, 10])/4 # Four motifs were reported

down5_list <- list()
down5_list[[paste0("Enriched in 5'-UTR sequences\nDown vs. Ref., p = ", pval)]] <- down5
plogo_down5 <- ggseqlogo(down5_list) + theme_logo(base_size = 10)
```
```{r}
# Motif enriched in 3'-UTR sequences of TE Up group vs Reference
up3_res <- read.table("./STREME/streme_TE_up-vs-Ref_seq_utr3_3-6nt.txt", skip = 29, fill = TRUE, flush = TRUE)
up3 <- up3_res[3:8, 1:4]
up3 <- apply(up3, 1, as.numeric)
rownames(up3) <- c("A", "C", "G", "U")
pval <- as.numeric(up3_res[2, 10])/4

up3_list <- list()
up3_list[[paste0("Enriched in 3'-UTR sequences\nUp vs. Ref., p = ", pval)]] <- up3
plogo_up3 <- ggseqlogo(up3_list) + theme_logo(base_size = 10)
```

#### Nucleotide identities near 5'cap, start codon

Chi-square analysis to compare proportion and calculate ratio change for Up or Down vs Ref.
```{r}
compare_chisq <- function(udf, var = "nt_p04") {
  df <- with(udf, table(TE_change_PPvsP, get(var)))
  chires <- pairwiseNominalIndependence(df, method = "BH")
  res <- data.frame(df)
  res <- res %>% group_by(TE_change_PPvsP) %>% mutate(fraction = Freq/sum(Freq))
  ref <- res[which(res$TE_change_PPvsP == "Reference"), ]
  res <- left_join(res, ref[, c("Var2", "fraction")], by = "Var2", suffix = c("", "_Reference"))
  res$fraction_ratio <- res$fraction / res$fraction_Reference
  res$log2_fraction_ratio <- log2(res$fraction_ratio)
  res$chisq_padj <- NA
  res[which(res$TE_change_PPvsP == "Up"), ]$chisq_padj <- chires[which(chires$Comparison == "Reference : Up"), ]$p.adj.Chisq
  res[which(res$TE_change_PPvsP == "Down"), ]$chisq_padj <- chires[which(chires$Comparison == "Reference : Down"), ]$p.adj.Chisq
  res$chisq_padj_UPvsDown <- chires[which(chires$Comparison == "Up : Down"), ]$p.adj.Chisq
  return(res)
}
```
```{r}
nts <- colnames(udf)[grepl(pattern = "^nt_.*[0-9][0-9]$", x = colnames(udf))]
ntdf_res <- lapply(nts, function(x) compare_chisq(udf, var = x))
names(ntdf_res) <- nts
ntdf <- bind_rows(ntdf_res, .id = "position")
ntdf <- ntdf[!is.na(ntdf$chisq_padj), ]
ntdf$Var2 <- sub("T", "U", ntdf$Var2)
ntdf$sign <- ifelse(test = grepl(pattern = "_m", x = ntdf$position), yes = "-", no = "+")
ntdf$num <- as.numeric(sub("nt_[a-z]?", "", ntdf$position))
ntdf[grepl(pattern = "_p", x = ntdf$position), ]$num <- ntdf[grepl(pattern = "_p", x = ntdf$position), ]$num + 1 # In the analysis, AUG is considered 0 +1 +2, so the nucloetide positions after it have to be fixed for plotting and the convension of AUG being +1 +2 +3
ntdf$pos <- paste(ntdf$sign, ntdf$num)
ntdf$from <- ifelse(test = grepl(pattern = "_[p|m]", x = ntdf$position), yes = "start", no = "cap")
ntdf$sig <- ifelse(test = ntdf$chisq_padj < 0.05, yes = "p < 0.05", no = "ns")
ntdf$sig <- factor(ntdf$sig, levels = c("p < 0.05", "ns"))
ntdf$hw <- ifelse(test = ntdf$sig != "ns", yes = 0.9, no = 0.6)
ntdf$sig_UpvsDown <- ifelse(test = ntdf$chisq_padj_UPvsDown < 0.05, yes = "p < 0.05", no = "ns")
ntdf$sig_UpvsDown <- factor(ntdf$sig_UpvsDown, levels = c("p < 0.05", "ns"))
```

Plot
```{r}
plot_compare_chisq <- function(ntdf, pos_to_plot, lims = c(-1.2, 1.2), xlabel = "") {
  ntdf2 <- ntdf[ntdf$position %in% pos_to_plot, ]
  levs <- unique(ntdf[ntdf$position %in% pos_to_plot, c("position", "pos")])
  levs <- levs[match(pos_to_plot, levs$position), ]
  levs <- levs$pos
  ntdf2$pos <- factor(ntdf2$pos, levels = levs)
  p <- ggplot(ntdf2, 
              aes(x = Var2, fill = log2_fraction_ratio, y = TE_change_PPvsP)) + 
    geom_tile(aes(width = hw, height = hw, linewidth = sig)) +
    geom_point(aes(alpha = sig_UpvsDown), shape = 8, size = 1) + 
    scale_fill_gradient2(name = expression(paste(log[2], " relative proportion to Reference")), 
                         low = "blue", mid = "white", high = "red", midpoint = 0, limits = lims) +
    scale_linewidth_manual(name = "Group vs. Reference:", values = c(`p < 0.05` = 0, ns = 3)) +
    scale_alpha_manual(name = "Up vs. Down:", values = c(`p < 0.05` = 1, ns = 0)) +
    scale_y_discrete(limits = rev) +
    facet_grid(.~pos, scales = "free_x", space = "free_x") + xlab(xlabel) + ylab("") +
    theme_bw(base_size = 10) + 
    theme(panel.grid = element_blank(), 
          legend.position = "top", legend.box = "horizontal", legend.title.align = 0.5,
          strip.background = element_rect(fill = "white"), panel.spacing = unit(0, "in")) +
    guides(fill = guide_colourbar(order = 1, barwidth = unit(5, "cm"), barheight = unit(0.25, "cm"), 
                                  title.position = "top"), 
           linewidth = guide_legend(order = 2, override.aes = list(color = "white"), 
                                    keywidth = unit(0.3, "cm"), keyheight = unit(0.3, "cm"),
                                    title.position = "top"),
           alpha = guide_legend(order = 3, keywidth = unit(0.3, "cm"), keyheight = unit(0.3, "cm"), 
                                title.position = "top"))
  return(p)
}
```
```{r}
# nt position after start codon
p_nts_p <- plot_compare_chisq(ntdf, pos_to_plot = nts[1:18], xlabel = "Nucleotide position relative to main ORF's AUG (+1 +2 +3)")
# nt position before start codon
p_nts_m <- plot_compare_chisq(ntdf, pos_to_plot = nts[36:19], xlabel = "Nucleotide position relative to main ORF's AUG (+1 +2 +3)")
# nt position from 5' cap
p_nts_5 <- plot_compare_chisq(ntdf, pos_to_plot = nts[37:54], xlabel = "Nucleotide position from 5' cap")
```

### Combine and export plots
```{r}
library(patchwork)
library(Cairo)
CairoFonts(
  regular = "Arial:style=Regular",
  bold = "Arial:style=Black",
  italic = "Arial:style=Italic",
  bolditalic = "Arial:style=Black Italic",
  symbol = "Symbol"
)
```

Figure 5 -- TE changes
```{r}
design <- "
ADD
BEE
CFF
"
fcA <- p_TE_main + theme(legend.title = element_blank(), legend.justification = c(1, 0.5))
fcf <- p_ripseq + theme(legend.position = "bottom")
fc_all <- (fcA + p_TE_pbp1d_mRNA + p_TE_mRNA_changes + p_riboseq + p_ripseq2 + p_ripseq) +
  plot_layout(design = design) +
  plot_annotation(tag_levels = "A") & theme(plot.tag = element_text(size = 12, face = "bold"), plot.tag.position = "topleft")
cairo_pdf(filename = "Fig5_TE_changes.pdf", family = "Arial", width = 7.5, height = 8.5) 
fc_all
dev.off()
```

Figure 6 -- TE and 5'end features
```{r}
design <- "
AABBBB
CCDDEE
"
f5E <- (plogo_down5 / plot_spacer())
f5_all <- (p5 + uorf$plot_res + ud5_A$plot_res + ud5_U$plot_res + f5E) +
  plot_layout(design = design) +
  plot_annotation(tag_levels = "A") & theme(plot.tag = element_text(size = 12, face = "bold"), plot.tag.position = "topleft")
cairo_pdf(filename = "Fig6_TE_5feats.pdf", family = "Arial", width = 7.5, height = 6.5) 
f5_all
dev.off()
```

Figure 7 -- TE and 3'end features
```{r}
design <- "
AB
CE
DF
G#
"
f3_all <- (pc + pm + p_re + p_ref + p3 + ud3_A$plot_res + plogo_up3) +
  plot_layout(design = design, heights = c(1, 1, 1, 0.5)) +
  plot_annotation(tag_levels = "A") & theme(plot.tag = element_text(size = 12, face = "bold"), plot.tag.position = "topleft")
cairo_pdf(filename = "Fig7_TE_3feats.pdf", family = "Arial", width = 5, height = 9.5) 
f3_all
dev.off()
```

Figure S7 -- TE and 5' nucleotide features
```{r}
p_ripseq_scatter2 <- p_ripseq_scatter + 
  theme(legend.position = "right") + 
  guides(color = guide_legend(reverse = TRUE, override.aes = list(alpha = 1)), fill = guide_legend(reverse = TRUE, override.aes = list(alpha = 1)))

uorf_cdf2 <- uorf_cdf + 
  theme(legend.position = "right")

p_nt_all <- (p_nts_m / p_nts_p / p_nts_5) + 
  plot_layout(guides = "collect") & theme(plot.tag = element_text(size = 12, face = "bold"), plot.tag.position = "topleft", legend.position = "top")

p_te_supp <- cowplot::plot_grid(p_ripseq_scatter2, uorf_cdf2, p_nt_all, 
                                ncol = 1, rel_heights = c(2.75, 2.2, 3.8), labels = "AUTO", label_size = 12, label_fontface = "bold")

cairo_pdf(filename = "FigS7.pdf", family = "Arial", width = 7.5, height = 8.75) 
p_te_supp
dev.off()
```

Figure S8 -- TE vs RIPseq vs transcript length
```{r}
cairo_pdf(filename = "FigS8_TE_RIPseq_transcriptlength.pdf", family = "Arial", width = 7.5, height = 3.75) 
p_ripseq3
dev.off()
```

